app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,241 @@
+                type: "application/x-tar",
+                disposition: "attachment"
+    else
+      send_file(absolute_path,
+                filename: File.basename(absolute_path),
+                disposition: 'attachment')
+    end
+  end
+
+  def upload_file(path)
+    absolute_path = check_path_exist(path)
+    if Archive.in_dir?(BASE_DIRECTORY, absolute_path, strict: false)
+      raise "You cannot upload files/create folders in the root directory click " \
+        "#{view_context.link_to 'here', new_course_url, method: 'get'}" \
+        " if you want to create a new course."
+    else
+      raise ActionController::ForbiddenError unless File.directory?(absolute_path)
+
+      if check_instructor(absolute_path) && !params[:name].nil?
+        all_filenames = Dir.entries(absolute_path)
+        if params[:name] != ""
+          if all_filenames.include?(params[:name].to_s)
+            raise "File with name #{input_file.original_filename} already exists."
+          end
+
+          # Creating a folder
+          dir = "#{absolute_path}/#{params[:name]}"
+          FileUtils.mkdir_p(dir)
+
+        else
+          # Uploading a file
+          input_file = params[:file]
+          return unless input_file
+          if all_filenames.include?(input_file.original_filename)
+            raise "File with name #{input_file.original_filename} already exists."
+          elsif input_file.size >= 1.gigabyte
+            raise "File size is too large. Upload a file that is smaller than 1 GB."
+          else
+            File.open(Rails.root.join(absolute_path, input_file.original_filename), 'wb') do |file|
+              file.write(input_file.read)
+            end
+          end
+        end
+      end
+    end
+  end
+
+  def my_escape(string)
+    string.gsub(/([^ a-zA-Z0-9_.-]+)/) do
+      "%#{$1.unpack('H2' * $1.bytesize).join('%').upcase}"
+    end
+  end
+
+  def populate_directory(current_directory, current_url)
+    directory = Dir.entries(current_directory)
+    new_url = current_url == '/' ? '' : current_url
+    @directory = directory.map do |file|
+      abs_path_str = "#{current_directory}/#{file}"
+      stat = File.stat(abs_path_str)
+      is_file = stat.file?
+      if %w[. ..].include?(file)
+        inst = true
+        if current_directory == BASE_DIRECTORY
+          inst = false
+        end
+      else
+        abs_path = Pathname.new(abs_path_str)
+        inst = check_instructor(abs_path)
+      end
+      {
+        size: (if is_file
+                 begin
+                   stat.size
+                 rescue StandardError
+                   '-'
+                 end
+               else
+                 '-'
+               end),
+        type: (is_file ? :file : :directory),
+        date: begin
+          stat.mtime.strftime('%d %b %Y %H:%M')
+        rescue StandardError
+          '-'
+        end,
+        relative: CGI.unescape(my_escape("/file_manager/#{new_url}#{file}")),
```

Comment:
Could you clarify what is the purpose of escaping and then unescaping the string?

Reply:
I changed it to just unescape.